### PR TITLE
Cross Server RPC Handler and Sender

### PR DESCRIPTION
**Pull request by jessicafalk**
#### Description
- Created a Cross Server RPC Handler and Sender
- We now only send a command response after executing the RPC
- enabling the command retry handler

TODO:
- write tests
- add function GUID into payload
- adding some error logs

#### Release note
wip
#### Tests
wip

#### Documentation
wip

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -1579,16 +1578,24 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
...
+	const FUnrealObjectRef CallingObjectRef = PackageMap->GetUnrealObjectRefFromObject(CallingObject);
 	if (!CallingObjectRef.IsValid())
 	{
 		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("The target object %s is unresolved; RPC %s will be dropped."),
 			   *CallingObject->GetFullName(), *Function->GetName());
 		return;
 	}
-	RPCPayload Payload = Sender->CreateRPCPayloadFromParams(CallingObject, CallingObjectRef, Function, Parameters);
 
-	Sender->ProcessOrQueueOutgoingRPC(CallingObjectRef, MoveTemp(Payload));
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1591]: What was the reasoning behind pulling the cross server RPC out of the sender? When we implement Cross Server RPC based on state, we might want them to follow the same path as the other RPCs (to get the queuing behaviour on overflow for instance).

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1591]: (what might happen is that the sender and handler you wrote get moved to the RPC service that Jacques has re-written.)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#include "Interop/GlobalStateManager.h"
+#include "Interop/SpatialPlayerSpawner.h"
+#include "Interop/SpatialReceiver.h"
+#include "Interop/SpatialSender.h"
+#include "Utils/RepLayoutUtils.h"
+#include "Utils/SpatialMetrics.h"
+
+using namespace SpatialGDK;
+
+CommandRequestResult RPCUtils::ExecuteCommand(const FCrossServerRPCParams& Params, USpatialNetDriver* NetDriver)
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCUtils.cpp:14]: This duplicates a lot of code from receiving RPCs in general, right ?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (jessicafalk) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1591]: yeah, and that is ok. At the moment cross server RPCs are not based on state and have a fairly different flow compared to all other RPCs, so it doesn't make much sense to have it in the RPC Service.
The reason it got pulled out of the Sender and the receiver is because the overall flow at the moment is hard to follow and reason about as there is just so much branching going on (e.g. it took me a while to identify that we actually never queue cross server RPCs).

So the sender and handler are mainly here to simplify the overall behaviour. when we then implement them based on state, we probably can get rid of this.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCUtils.cpp (hunk 2 above)
Comment (jessicafalk) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCUtils.cpp:14]: that is the idea behind the RPC Utils  :) the overall cleanup isn't completely done yet, because I wanted to wait for Jacques' PR to go in, otherwise we may run into conflicts.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (jacquese) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1591]: It could be worth moving creating payload logic (which seems to be shared between both cross server and not) into some utilities class of the RPC service, then move the cross server RPCs into the RPC service as well. Right now in the rewrite though, sending is not fully handled by the RPC service, so it probably doesn't make sense yet. 

For reference, I was thinking the RPC service would be more of a logical encapsulation of all things RPCs rather than unifying flows, so that there's one central place where a developer can go to reason about them when making changes/debugging. The multicast and client server ones are more similar to each other than to cross server RPCs, but still fairly different.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h~ [hunk 3]
```diff
@@ -0,0 +1,313 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriver.h~:1]: Looks like this was added by mistake.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialView/ViewDelta.cpp [hunk 4]
```diff
@@ -139,6 +139,11 @@ const TArray<EntityDelta>& ViewDelta::GetEntityDeltas() const
 	return EntityDeltas;
 }
 
+TArray<Worker_Op>& ViewDelta::GetWorkerMessages()
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialView/ViewDelta.cpp:142]: I do not see this non-const version used anywhere?

Comment (jessicafalk) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialView/ViewDelta.cpp:142]: ah this was a remnant of sami's work and he hadn't removed it. will remove it

**Comment by jessicafalk** [comment]
got closed automatically: decided to split the code out into 2 new PRs:
https://github.com/spatialos/UnrealGDK/pull/2665
https://github.com/spatialos/UnrealGDK/pull/2666
